tests/unit/test_lib.py

```diff
@@ -0,0 +1,139 @@
+# Copyright 2023 Canonical
+# See LICENSE file for licensing details.
+
+"""Tests for the pushgateway.py charm library."""
+
+import json
+from unittest.mock import patch
+
+import pytest
+import requests
+import responses
+from ops.testing import Harness
+
+from tests.testingcharm.src.charm import TestingcharmCharm
+
+
+@pytest.fixture()
+@patch("charm.KubernetesServicePatch", lambda x, y: None)
+def harness():
+    harness = Harness(TestingcharmCharm)
+    harness.begin()
+    return harness
+
+
+def test_pushgateway_url_default(harness):
+    """The pushgateway url has a default value."""
+    ppi = harness.charm.ppi
+    assert ppi._stored.pushgateway_url is None
+
+
+def test_pushgateway_relation_changed_empty(harness):
+    """No changes when the relation is established without data."""
+    ppi = harness.charm.ppi
+    harness.add_relation("pushgateway", "remote")
+    assert ppi._stored.pushgateway_url is None
+    assert not ppi.is_ready
+
+
+def test_pushgateway_relation_changed_with_data(harness):
+    """The pushgateway url is set when the relation is established and has data."""
+    ppi = harness.charm.ppi
+    payload = {"push-endpoint": json.dumps({"port": "9876", "hostname": "hostname.test"})}
+
+    relation_id = harness.add_relation("pushgateway", "remote")
+    harness.update_relation_data(relation_id, "remote", payload)
+    assert ppi._stored.pushgateway_url == "http://hostname.test:9876/"
+    assert ppi.is_ready
+
+
+def test_pushgateway_relation_broken(harness):
+    """The pushgateway url is cleared if the relation disappears."""
+    ppi = harness.charm.ppi
+    payload = {"push-endpoint": json.dumps({"port": "9876", "hostname": "hostname.test"})}
+    relation_id = harness.add_relation("pushgateway", "remote")
+    harness.update_relation_data(relation_id, "remote", payload)
+    assert ppi.is_ready
+
+    harness.remove_relation(relation_id)
+    assert not ppi.is_ready
+
+
+@pytest.mark.parametrize(
+    "name",
+    [
+        "moño",  # not ascii
+        123,  # not a number
+        "",  # empty
+    ],
+)
+@responses.activate
+def test_sendmetric_bad_name_input(harness, name):
```

Comment:
that's me, but I'd lump these three validation tests in one:

```
invalid = object()
@parametrize("key, value, expect", (
('foo', 3, 'foo 3'),
('foo', 3.4, 'foo 3.4'),
(42, 'foo', invalid),
))
...
```

feel free to ignore this one, it's a nit.

Reply:
I prefer to have them separated, for the error message.